Rework TreePath Cache

ChrisKujawa opened the pull request:
## PR Description

The PR is divided into three commits, first a bigger refactoring to extract the treePath cache and make it easier to test and maintain. As a second step the cache has been integrated into the existing FlowNodeInstanceProcessor, at the end the bug was fixed to add the missing flowScopeKey with corresponding treePath.


> [!Important]
>
> I wanted to keep the PR small, and stop here and keep it simple for now. So as not to overwhelm the reviewer with too many changes.
> 
> The next steps are:
> 
> 1. Add metrics for cache access and misses, metric for time of resolver
> 2. Clean up - keys don't need to be strings
> 3. Fixing the issue with adding unnecessary leafes to the cache https://github.com/camunda/camunda/issues/22076 
> 4. Adjust processor to not use treePath if element completes, terminates, etc https://github.com/camunda/camunda/issues/22075

### Refactoring

Create a new class for the treePath cache, FlowNodeInstanceTreePathCache.

The FlowNodeInstanceTreePathCache comprises the corresponding multi-level HashMap (with SoftHashMap), and the resolvment via a given treePathResolver.

A new record was created to encapsulate the necessary values like, partitionId, recordKey, flowScopeKey, and processInstanceKey. This allows us to reduce the dependency to the Zeebe protocol. Allows us to move the cache to the common importer module.

As the resolver is available via DI the treePath cache can be easily tested via unit tests.

New tests and documentation were added to cover and describe the behavior of the cache.

### Integration

The integration was rather simple and allowed to remove several code/complexity from the processor.

### Fix

Previously when the treePath of the flowScope was not part of the cache the resolver was used to resolve the treePath. This was done every time as the resolved treePath was not added to the cache. Causing unnecessary multiple calls to the resolver.
<!-- Describe the goal and purpose of this PR. -->

## Related issues

closes https://github.com/camunda/camunda/issues/22074

ChrisKujawa committed:
refactor: extract treePath cache

Create a new class for the treePath cache,
FlowNodeInstanceTreePathCache.

The FlowNodeInstanceTreePathCache comprises the corresponding
multi-level HashMap (with SoftHashMap), and the resolvment via a given
treePathResolver.

A new record was created to encapsulate the necessary values like,
partitionId, recordKey, flowScopeKey and processInstanceKey. This
allows to reduce the dependency to the Zeebe protocol. Allows us to move
the cache to the common module.

As the resolver is available via DI the cache can be easily tested via
unit-tests.

Added new tests and documentation to cover and describe the behavior of
the cache.

ChrisKujawa committed:
refactor: integrate new cache

Replace old cache logic with new FlowNodeInstanceTreePathCache

ChrisKujawa committed:
fix: do not resolve treePath twice

Previously the flowScopeKey and corresponding treePath was not added to the cache
after it was resolved via the resolver, this caused to
unnecessarily multiple resolvements

Zelldon commented:
I target `[operate-s-22328-cache-partitions](https://github.com/camunda/camunda/tree/operate-s-22328-cache-partitions)` as I think @ralfpuchert is working on this branch to get it merged into stable. After this the based will change to stable/8.5.

ChrisKujawa committed:
refactor: rename FlowNodeInstanceRecord

Review suggestion to rename class to FNITreePathCacheCompositeKey
as it is a compositeKey used in the treePath, to avoid confusion with
other record classes and FNI classes

ChrisKujawa committed:
refactor: clarify treePathResolver in tests

Zelldon commented:
@ThorbenLindhauer I removed the accidentially added files

Furthermore, renamed the FlowNodeInstanceRecord class https://github.com/camunda/camunda/pull/22120/commits/378a4d16fd5a35818cac946f115e52f7c7696112

Refactored the resolver variable name and added some doc to clarify https://github.com/camunda/camunda/pull/22120/commits/4a53fa51767960eb8a7a4ecf5018bfc46a942ead

Let me know whether there are more changes necessary. 

What is the state of merging the `operate-s-22328-cache-partitions` ?

ralfpuchert reviewed (approved):
LGTM! This PR makes the [PR](https://github.com/camunda/camunda/pull/21059) obsolete. It solves the same issue better imo. It is more clear, separated and tested.

We can merge it into `stable/operate-8.5`, can't we?

romansmirnov commented on operate/importer-common/src/main/java/io/camunda/operate/zeebeimport/cache/FlowNodeInstanceTreePathCache.java:
```diff
@@ -0,0 +1,144 @@
+              "Unable to find parent tree path for flow node instance id [{}], parent flow node instance id [{}]",
+              compositeKey.recordKey(),
+              compositeKey.flowScopeKey());
+          parentTreePath = ConversionUtils.toStringOrNull(compositeKey.processInstanceKey());
```
Shouldn't this be put into the cache as well? (I mean, if there is no `treePath`, then the next one doesn't need to do a search request.)

Zelldon replied:
Here I was not sure, as I was also not sure whether the error handling itself is correct tbh. Because we didn;t found the flowScope and just assume it is the PI 🤔

Zelldon replied:
As this is the same as before I don't want to change as part of this PR. Happy to do it in an additional PR when we have clarified whether this is correct or not 😅

Zelldon commented:
> LGTM! This PR makes the [PR](https://github.com/camunda/camunda/pull/21059) obsolete. It solves the same issue better imo. It is more clear, separated and tested.
> 
> We can merge it into `stable/operate-8.5`, can't we?

Fine by me. I didn't wanted to overrule your changes :) so I was working based on your branch here `operate-s-22328-cache-partitions`.

I can change the base/target branch if you want.

I guess I need to change also the flow-node 8.4 processor then?

Zelldon commented:
I think we can merge https://github.com/camunda/camunda/pull/21059 to stable (as this lays the foundation) and then I resolve the conflicts and we merge this PR here to stable :)

romansmirnov reviewed (commented):
Just a small comment.

ChrisKujawa committed:
refactor(8.4): integrate new cache

Replace old cache logic with new FlowNodeInstanceTreePathCache
on the 8.4 importer

Zelldon commented:
Hey @ralfpuchert @ThorbenLindhauer I rebased the PR now to stable/operate-8.5 and also made changes on the [8.4 importer](https://github.com/camunda/camunda/pull/22120/commits/3eb44a4a15eb7fd2fa79f390e6e24e4d1abc2955)

Let me know whether we can merge the PR.

ralfpuchert commented:
@Zelldon, reviewed changes in importer for 8.4. LGTM.

Zelldon commented:
/backport

backport-action commented:
Created backport PR for `main`:
- #22203 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-22120-to-main
git worktree add --checkout .worktree/backport-22120-to-main backport-22120-to-main
cd .worktree/backport-22120-to-main
git reset --hard HEAD^
git cherry-pick -x 199281b9c297c2beb8bf81b95fc13839ae3ab09d 204300d4ff910bf58ebbc90cf488e8a9d10d5007 5fdd59e481cc5de5dd14609efc360f7adc39f3a4 24468934d235755a259beedaa21386b79dae0128 3eb44a4a15eb7fd2fa79f390e6e24e4d1abc2955
git push --force-with-lease
```
